automated-testing/synthetic-monitoring-tests/readme.md

```diff
@@ -0,0 +1,82 @@
+It is not uncommon for the same code for e2e or integration tests to be used to implement the probe. Probes target front-end or publicly-facing APIs.
+
+### Monitoring
+
+Given that Synthetic Monitoring tests are continuously running, at intervals, in a production environment, the assertion of system behavior through analysis relies on existing monitoring pillars used in live system (Logging, Metrics, Distributed Tracing).
+There would usually be a finite set of tests, and key metrics that are used to build monitors and alerts to assert against the known [SLO](https://en.wikipedia.org/wiki/Service-level_objective), and verify that the [OKR](https://en.wikipedia.org/wiki/OKR) for that system are maintained. The monitoring tools are effectively capturing both RUMs and synthetic data generated by the probes.
+
+## Applying Synthetic Monitoring Tests [the how]
+
+### Asserting the system under tests
+
+Synthetic monitoring tests are usually statistical. Test metrics are compared against some historical or running average with a time dimension *(Example: Over the last 30 days, for this time of day, the mean average response time is 250ms for AddToCart operation with a standard deviation from the mean of +/- 32ms)*. So if an observed measurement is within a [deviation of the norm](https://en.wikipedia.org/wiki/Standard_deviation) at any time, the services are probably healthy.
+
+### Monitoring the health of tests
+
+Probes runtime is a production environment on its own, and the health of tests is critical. Many providers offer cloud-based systems that host such runtimes, while some organizations use existing production environments to run these tests on. In either way, a monitor-the-monitor strategy should be a first-class citizen of the production environment's alerting systems.
+
+### Synthetic Monitoring and Real User Monitoring
+
+Synthetic monitoring does not replace the need for RUM. Probes are predictable code that verifies specific scenarios, and they do not 100% completely and truly represent how a user session is handled. On the other hand, prefer not to use RUMs to test for site reliability because:
+
+* As the name implies, RUM requires user traffic. The site may be down, but since no user visited the monitored path, no alerts were triggered yet.
+* Inconsistent Traffic and usage patterns make it hard to gauge for benchmarks.
+
+### Risks
+
+Testing in production, in general, has a risk factor attached to it, which does not exist tests executed during CI/CD stages. Specifically, in synthetic monitoring tests, the following may affect the production environment:
+
+* Corrupted or invalid data - Tests inject test data which may be in some way corrupt. Consider using a testing schema.
+* Protected data leakage - Tests run in a production environment and emit logs or trace that may contain protected data.
+* Overloaded systems - Synthetic tests may cause errors or overload the system.
```

Comment:
:-) I once owned a system that over time accumulated so much test-originated data that the test cases failed because the performance degraded, even tho real users' traffic was fine -- so the synthetic transactions set off a false positive alert that the system was unhealthy.  Then we had to build a garbage collector to fix and prevent the problem...

Reply:
:)
would you like me to add something here or can I resolve?

Reply:
pls resolve.  i couldn't resist telling the story, sorry.